### ISSUE TITLE
orsopy test suite fails on cp312.

### ISSUE THREAD

**Issue by andyfaff** [reflectivity/orsopy#115]
The test suite fails on cp312, as mentioned in #114.

This triggers the fail:
```
from orsopy.fileio.base import ErrorValue
```

The cause behind it is the `dataclass` package changing it's internals, which we rely on. Our usage of dataclasses seems fragile - we monkeypatch the internals (which are marked private) to permit extra kwargs when instantiating classes.

Given this fragility, I'm wondering if it's better to stop using dataclasses and/or define customised `__init__` methods for all our Header-derived classes.

**Comment by andyfaff** [comment]
Or do something like:

```
@dataclass
class A:
    a: int
    kwargs: dict=field(default_factory=dict)

    def __post_init__(self, *args):
        for k, v in self.kwargs.items():
            setattr(self, k, v)
```

**Comment by bmaranville** [comment]
If user kwargs are split out this way on class instantiation we'll have to do something similar when serializing, so that all non-orso attrs are stuffed back into a kwargs dict, so that these structures can survive a serialize/deserialize round trip.

**Comment by rolandmas** [comment]
For the record: this causes Debian bug https://bugs.debian.org/1058435

**Comment by bmaranville** [comment]
That is surprising - who is packaging orsopy for Debian release?  Do we have documentation on that process, who is running it, and what the workflows are?

**Comment by rolandmas** [comment]
I am :-)
The workflow is the standard Debian workflow: someone (the packager/maintainer) picks the "upstream" releases, adds what's needed for packaging, patches if needed, ensures it works well with the rest of Debian, then uploads the package to the "unstable" repository. Then the package follows its course, first through "testing", then to the next "stable" release of Debian if it fulfills a few criteria (no critical bugs reported or left open for some time).
The maintainer also acts as a link between Debian users and the upstream authors, transmitting bug reports and fixes across. Which is what I just did :-)

**Comment by bmaranville** [comment]
Oh, and thanks for the contributions!  Sorry I forgot to mention that :)

**Comment by rolandmas** [comment]
There is pretty extensive documentation on the Debian wiki, for instance starting with https://wiki.debian.org/DebianDevelopment
There's a team of Debian developers who focus on Python-related packages, and they decide collaboratively which version of Python to support. For 3.12, it's been tested outside of Debian repositories (or in the Debian "experimental" repository) for a while, which resulted in a lot of bug reports; when most of these were fixed, Python 3.12 has been injected into Debian proper, in order to give more visibility to the remaining problems. Some time before the actual release of the next Debian stable distribution, they'll consider which Python version gets to be the default, and whether other versions are also provided alongside.
Note that for testing, there are also a bunch of automated testing systems in Debian. Some results are automatically listed on https://tracker.debian.org/pkg/orsopy 
Note also that the actual packaging happens on https://salsa.debian.org/python-team/packages/orsopy (you may want to have a look at the debian/patches subdirectory)
Cheers!

**Comment by andyfaff** [comment]
refnx is developed as follows.

I regularly release point versions with bug fixes/updates/new features (it's lightweight to make a release). refnx uses compiled code so each version has a build matrix over a range of Python versions.

For a given refnx version the oldest python that's supported (i.e. tested in CI, specified in pyproject.toml, and built) is similar to the python version supported by the most recent version of scipy/numpy, which use [SPEC0/NEP29](https://scientific-python.org/specs/spec-0000/) for which python versions they support. This means the latest version of refnx is specified to only use Python > 3.9.

The most recent Python version that's supported would be a Python for which there is a numpy/scipy/cython version available on PyPI. Sometimes this is a pre-release Python. refnx has no hard dependency on other packages. It does have to install orsopy, but that's pure Python.
It hurts if orsopy does not work with cp312, because many users will use the most recent Python version. Indeed, cp312 is the default Python installed by conda.

It's good practice for many projects to have a CI entry that runs against bleeding edge upstream dependencies, both e.g. numpy, Python, so that one gets good notice of when a major release is likely to break something.

**Comment by aglavic** [comment]
Resolved by #120
